### PR TITLE
[KYUUBI #544] Deregister engine when meeting specifies exception messages

### PR DESCRIPTION
**Pull request by turboFei**
### _Why are the changes needed?_

Involve `kyuubi.engine.deregister.exception.messages` to deregister engine when meeting specified exception messages.

### _How was this patch tested?_
Existing UT.
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.readthedocs.io/en/latest/tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by pan3793** [comment]
Could you please rebase on master to resolve the conflicts?

**Comment by turboFei** [comment]
sure, rebasing

**Review by pan3793 — APPROVED** [review]
LGTM, thanks!

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala [hunk 1]
```diff
@@ -41,15 +42,28 @@ class SparkSQLEngineListener(server: Serverable) extends SparkListener with Logg
...
+     case JobFailed(e) if e != null =>
+       var shouldDeregister = false
+       if (deregisterExceptions.exists(_.equals(e.getClass.getCanonicalName))) {
          error(
            s"""
              | Job failed exception class is in the set of
              | ${ENGINE_DEREGISTER_EXCEPTION_CLASSES.key}, stopping the engine.
            """.stripMargin, e)
+         shouldDeregister = true
+       } else if (deregisterMessages.exists(e.getMessage.contains)) {
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala:54]: NPE？

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala [hunk 2]
```diff
@@ -21,13 +21,14 @@ package org.apache.spark.kyuubi
...
 import org.apache.kyuubi.Logging
-import org.apache.kyuubi.config.KyuubiConf.ENGINE_DEREGISTER_EXCEPTION_CLASSES
+import org.apache.kyuubi.config.KyuubiConf.{ENGINE_DEREGISTER_EXCEPTION_CLASSES, ENGINE_DEREGISTER_EXCEPTION_MESSAGES}
 import org.apache.kyuubi.ha.client.EngineServiceDiscovery
 import org.apache.kyuubi.service.{Serverable, ServiceState}
 
 class SparkSQLEngineListener(server: Serverable) extends SparkListener with Logging {
 
   lazy val deregisterExceptions = server.getConf.get(ENGINE_DEREGISTER_EXCEPTION_CLASSES)
+  lazy val deregisterMessages = server.getConf.get(ENGINE_DEREGISTER_EXCEPTION_MESSAGES)
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala:31]: Add some comments why these variable should be lazy

File: externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala [hunk 3]
```diff
@@ -41,15 +42,28 @@ class SparkSQLEngineListener(server: Serverable) extends SparkListener with Logg
...
            """.stripMargin, e)
+         shouldDeregister = true
+       } else if (deregisterMessages.exists(e.getMessage.contains)) {
+         error(
+           s"""
+             | Job failed exception message matches the specified
+             | ${ENGINE_DEREGISTER_EXCEPTION_MESSAGES.key}, stopping the engine.
+           """.stripMargin, e)
+         shouldDeregister = true
+       }
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala:61]: these codes duplicate w/ above

Comment (turboFei) [reply externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLEngineListener.scala]: extracted deregister info

**Review by yaooqinn — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks for @turboFei work and @yaooqinn reviewing, as all comments have been resolved, I will merge it to master for v1.2.0

**Pull request closed without merging**